### PR TITLE
fix(naming): Rename variant to implementation

### LINKED ISSUE #3: rename variant to implementation
The current code refers to "variants" in a number of places, typically when talking about multiple implementations of the same function (e.g. add<int32,int32> and add<int64,int64>).  Since the spec calls these "implementations" and the spec also has a different concept that it calls "variant" I think it would be good to update the repo changing all instances of "variant" to "implementation".

### PR DESCRIPTION
**Pull request by chaojun-zhang**
Closes #3

### CODE REVIEW

File: include/substrait/function/Function.h [hunk 1]
```diff
@@ -95,7 +95,7 @@ struct FunctionImplementation {
   ParameterizedTypePtr returnType;
   std::optional<FunctionVariadic> variadic;
 
-  /// Test if the actual types matched with this function variant.
+  /// Test if the actual types matched with this function implement.
```
Comment (EpsilonPrime) [inline include/substrait/function/Function.h:98]: It's longer but would this read better if it said "this function's implementation"?

**Review by EpsilonPrime — APPROVED** [review]

File: include/substrait/function/Function.h (hunk 1 above)
Comment (chaojun-zhang) [reply include/substrait/function/Function.h:98]: Rebased and updated

File: include/substrait/function/Extension.h [hunk 2]
```diff
@@ -40,40 +40,40 @@ class Extension {
...
       const std::vector<std::string>& extensionFiles);
 
-  /// Add a scalar function variant.
-  void addScalarFunctionVariant(const FunctionImplementationPtr& functionVariant);
+  /// Add a scalar function implementation.
+  void addScalarFunctionImpl(const FunctionImplementationPtr& functionImpl);
 
-  /// Add a aggregate function variant.
-  void addAggregateFunctionVariant(const FunctionImplementationPtr& functionVariant);
+  /// Add a aggregate function implementation.
```
Comment (westonpace) [inline include/substrait/function/Extension.h:46]:
```suggestion
  /// Add an aggregate function implementation.
```

File: include/substrait/function/Function.h (hunk 1 above)
Comment (westonpace) [inline include/substrait/function/Function.h:98]:
```suggestion
  /// Test if the actual types matched with this function's implementation.
```

**Review by westonpace — APPROVED** [review]
This is great, thanks.  Two minor wording nits (I can apply them myself and will merge).

**Pull request merged**